### ISSUE TITLE
JDK19 s390x serviceability_jvmti_j9_0_FAILED ASSERTION FAILED at vm/VMAccess.cpp:355: (0 == ((vmThread)->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS))

### ISSUE THREAD

**Issue by JasonFengJ9** [eclipse-openj9/openj9#16752]
Failure link
------------

From [an internal build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk19_j9_extended.openjdk_s390x_linux/38/)(`rhel7s390x-4-5`):
```
openjdk version "19.0.2" 2023-01-17
IBM Semeru Runtime Open Edition 19.0.2+7 (build 19.0.2+7)
Eclipse OpenJ9 VM 19.0.2+7 (build master-042864b42, JRE 19 Linux s390x-64-Bit Compressed References 20230117_46 (JIT enabled, AOT enabled)
OpenJ9   - 042864b42
OMR      - b86560818
JCL      - edc950934d based on jdk-19.0.2+7)
```
[Rerun in Grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/parambuild/?SDK_RESOURCE=customized&amp;TARGET=serviceability_jvmti_j9_0&amp;TEST_FLAG=&amp;UPSTREAM_TEST_JOB_NAME=&amp;DOCKER_REQUIRED=false&amp;ACTIVE_NODE_TIMEOUT=0&amp;VENDOR_TEST_DIRS=&amp;EXTRA_DOCKER_ARGS=&amp;TKG_OWNER_BRANCH=adoptium%3Amaster&amp;OPENJ9_SYSTEMTEST_OWNER_BRANCH=eclipse%3Amaster&amp;PLATFORM=s390x_linux&amp;GENERATE_JOBS=true&amp;KEEP_REPORTDIR=true&amp;PERSONAL_BUILD=false&amp;DOCKER_REGISTRY_DIR=&amp;ADOPTOPENJDK_REPO=https%3A%2F%2Fgithub.com%2Fadoptium%2Faqa-tests.git&amp;RERUN_ITERATIONS=3&amp;SETUP_JCK_RUN=false&amp;DOCKER_REGISTRY_URL_CREDENTIAL_ID=&amp;LABEL=&amp;EXTRA_OPTIONS=&amp;CUSTOMIZED_SDK_URL=+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk19u%2Fjdk19u-linux-s390x-openj9%2F46%2Fibm-semeru-open-jre_s390x_linux_JDK19U_2023-02-18-17-00.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk19u%2Fjdk19u-linux-s390x-openj9%2F46%2Fibm-semeru-open-testimage_s390x_linux_JDK19U_2023-02-18-17-00.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk19u%2Fjdk19u-linux-s390x-openj9%2F46%2Fibm-semeru-open-jdk_s390x_linux_JDK19U_2023-02-18-17-00.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk19u%2Fjdk19u-linux-s390x-openj9%2F46%2Fibm-semeru-open-debugimage_s390x_linux_JDK19U_2023-02-18-17-00.tar.gz&amp;BUILD_IDENTIFIER=&amp;JENKINS_KEY=&amp;ADOPTOPENJDK_BRANCH=master&amp;LIGHT_WEIGHT_CHECKOUT=false&amp;USE_JRE=false&amp;ARTIFACTORY_SERVER=na.artifactory.swg-devops&amp;KEEP_WORKSPACE=false&amp;USER_CREDENTIALS_ID=83181e25-eea4-4f55-8b3e-e79615733226&amp;JDK_VERSION=19&amp;DOCKER_REGISTRY_URL=&amp;ITERATIONS=1&amp;VENDOR_TEST_REPOS=&amp;JDK_REPO=https%3A%2F%2Fgithub.com%2Fibmruntimes%2Fopenj9-openjdk-jdk19&amp;JCK_GIT_BRANCH=main&amp;RELEASE_TAG=&amp;OPENJ9_BRANCH=master&amp;OPENJ9_SHA=&amp;JCK_GIT_REPO=&amp;VENDOR_TEST_BRANCHES=&amp;OPENJ9_REPO=https%3A%2F%2Fgithub.com%2Feclipse-openj9%2Fopenj9.git&amp;UPSTREAM_JOB_NAME=&amp;CLOUD_PROVIDER=&amp;CUSTOM_TARGET=&amp;VENDOR_TEST_SHAS=&amp;JDK_BRANCH=openj9&amp;LABEL_ADDITION=&amp;ARTIFACTORY_REPO=&amp;ARTIFACTORY_ROOT_DIR=&amp;UPSTREAM_TEST_JOB_NUMBER=&amp;DOCKERIMAGE_TAG=&amp;JDK_IMPL=openj9&amp;TEST_TIME=&amp;SSH_AGENT_CREDENTIAL=83181e25-eea4-4f55-8b3e-e79615733226&amp;AUTO_DETECT=true&amp;SLACK_CHANNEL=%23rt-jenkins&amp;DYNAMIC_COMPILE=false&amp;RELATED_NODES=&amp;ADOPTOPENJDK_SYSTEMTEST_OWNER_BRANCH=adoptium%3Amaster&amp;APPLICATION_OPTIONS=&amp;CUSTOMIZED_SDK_URL_CREDENTIAL_ID=4e18ffe7-b1b1-4272-9979-99769b68bcc2&amp;ARCHIVE_TEST_RESULTS=false&amp;NUM_MACHINES=&amp;OPENJDK_SHA=&amp;TRSS_URL=http%3A%2F%2Ftrss1.fyre.ibm.com&amp;USE_TESTENV_PROPERTIES=false&amp;BUILD_LIST=openjdk&amp;UPSTREAM_JOB_NUMBER=&amp;STF_OWNER_BRANCH=adoptium%3Amaster&amp;TIME_LIMIT=20&amp;JVM_OPTIONS=&amp;PARALLEL=None) - Change TARGET to run only the failed test targets.

Optional info
-------------

Failure output (captured from console output)
---------------------------------------------
```
[2023-02-18T18:30:33.726Z] variation: Mode150
[2023-02-18T18:30:33.726Z] JVM_OPTIONS:  -XX:+UseCompressedOops 

[2023-02-18T18:34:10.002Z] TEST: serviceability/jvmti/vthread/VThreadUnsupportedTest/VThreadUnsupportedTest.java

[2023-02-18T18:34:10.003Z] STDERR:
[2023-02-18T18:34:10.003Z] 18:33:50.645 0x28c600    j9vm.228    *   ** ASSERTION FAILED ** at /home/jenkins/workspace/build-scripts/jobs/jdk19u/jdk19u-linux-s390x-openj9/workspace/build/src/openj9/runtime/vm/VMAccess.cpp:355: (0 == ((vmThread)->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS))
[2023-02-18T18:34:10.003Z] JVMDUMP039I Processing dump event "traceassert", detail "" at 2023/02/18 10:33:50 - please wait.

[2023-02-18T18:34:22.626Z] serviceability_jvmti_j9_0_FAILED
```

Reproduced at [3x grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/1/consoleFull)

**Comment by tajila** [comment]
@JasonFengJ9 does this only fail on Z?

**Comment by tajila** [comment]
@r30shah ^^

**Comment by r30shah** [comment]
I tried running the failing job with `-Xint` and seems like it fails there as well ([internal link](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/3/). 
I am trying to confirm if the grinder did take `-Xint` or not (@JasonFengJ9  Can you confirm if the grinder is correct or not? )

**Comment by JasonFengJ9** [comment]
@r30shah your grinder did take `-Xint` as per `aqa-tests/TKG/output_16775976884901/serviceability_jvmti_j9_0/work/serviceability/jvmti/vthread/VThreadUnsupportedTest/VThreadUnsupportedTest/javacore.20230228.072451.3948.0002.txt` from the [result file](https://na.artifactory.swg-devops.com/artifactory/sys-rt-generic-local/hyc-runtimes-jenkins.swg-devops.com/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/3/openjdk_test_output.tar.gz).
```
1CICMDLINE     /home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/openjdkbinary/j2sdk-image/bin/java -Dtest.vm.opts=-ea -esa -Xmx512m -Xint -Dtest.tool.vm.opts=-J-ea -J-esa -J-Xmx512m -J-Xint -Dtest.compiler.opts= -Dtest.java.opts= -Dtest.jdk=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/openjdkbinary/j2sdk-image -Dcompile.jdk=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/openjdkbinary/j2sdk-image -Dtest.timeout.factor=8.0 -Dtest.nativepath=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/openjdkbinary/openjdk-test-image/hotspot/jtreg/native -Dtest.root=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/aqa-tests/openjdk/openjdk-jdk/test/hotspot/jtreg -Dtest.name=serviceability/jvmti/vthread/VThreadUnsupportedTest/VThreadUnsupportedTest.java -Dtest.file=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/aqa-tests/openjdk/openjdk-jdk/test/hotspot/jtreg/serviceability/jvmti/vthread/VThreadUnsupportedTest/VThreadUnsupportedTest.java -Dtest.src=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/aqa-tests/openjdk/openjdk-jdk/test/hotspot/jtreg/serviceability/jvmti/vthread/VThreadUnsupportedTest -Dtest.src.path=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/aqa-tests/openjdk/openjdk-jdk/test/hotspot/jtreg/serviceability/jvmti/vthread/VThreadUnsupportedTest -Dtest.classes=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/aqa-tests/TKG/output_16775976884901/serviceability_jvmti_j9_0/work/classes/1/serviceability/jvmti/vthread/VThreadUnsupportedTest/VThreadUnsupportedTest.d -Dtest.class.path=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/aqa-tests/TKG/output_16775976884901/serviceability_jvmti_j9_0/work/classes/1/serviceability/jvmti/vthread/VThreadUnsupportedTest/VThreadUnsupportedTest.d -Dtest.class.path.prefix=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/aqa-tests/TKG/output_16775976884901/serviceability_jvmti_j9_0/work/classes/1/serviceability/jvmti/vthread/VThreadUnsupportedTest/VThreadUnsupportedTest.d:/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/aqa-tests/openjdk/openjdk-jdk/test/hotspot/jtreg/serviceability/jvmti/vthread/VThreadUnsupportedTest -ea -esa -Xmx512m -Xint -Djava.library.path=/home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/openjdkbinary/openjdk-test-image/hotspot/jtreg/native --enable-preview -agentlib:VThreadUnsupportedTest com.sun.javatest.regtest.agent.MainWrapper /home/jenkins/workspace/Test_openjdk19_j9_extended.openjdk_s390x_linux_rerun/aqa-tests/TKG/output_16775976884901/serviceability_jvmti_j9_0/work/serviceability/jvmti/vthread/VThreadUnsupportedTest/VThreadUnsupportedTest.d/main.0.jta
```

Adding `vm` label for now.

**Comment by pshipton** [comment]
@tajila this is back to VM.

**Comment by pshipton** [comment]
@r30shah For the future, I expect you should set EXTRA_OPTIONS to add -Xint, and not JVM_OPTIONS to replace all other options with -Xint.

**Comment by tajila** [comment]
@babsingh please take a look

**Comment by babsingh** [comment]
In the test, `jvmtiHookVirtualThreadMount` invokes `jni->IsVirtualThread` (from the `callback`). `jvmtiHookVirtualThreadMount` does not release VM access and `jni->IsVirtualThread` tries to acquire VM access which causes the assertion.

**FIX** : `jvmtiHookVirtualThread[Mount|Unmount]` need to release VM access (`internalExitVMToJNI`) before invoking the `callback`. This should stop the assertion. Will implement and verify the fix in the next few days.

`jvmtiHookVirtualThread[Mount|Unmount]` are JVMTI extension events and not documented in the JVMTI spec. They are not needed for the JDK19 release. Re-targeting this issue to the next release.

**Comment by JasonFengJ9** [comment]
This failure appeared at [JDK19 milestone 1 build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk19_j9_extended.openjdk_s390x_linux/40/).

**Comment by babsingh** [comment]
> This failure appeared at [JDK19 milestone 1 build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk19_j9_extended.openjdk_s390x_linux/40/).

https://github.com/eclipse-openj9/openj9/pull/16816 was merged in the HEAD branch yesterday. https://github.com/eclipse-openj9/openj9/pull/16823 has been opened to deliver the fix to the 0.37 release branch. The above issue will be resolved once #16823 is merged.

**Comment by pshipton** [comment]
Since we've made a fix to 0.37 to fix this, I'll put it back in the 0.37 milestone plan. Since the fixed are merged I assume this can be closed?

**Comment by babsingh** [comment]
Yes, it can be closed.
